source/common/upstream/load_balancer_impl.cc

```diff
@@ -138,31 +162,58 @@ void LoadBalancerBase::recalculatePerPriorityState(uint32_t priority,
     // Everything is terrible. Send all load to P=0.
     // In this one case sumEntries(per_priority_load) != 100 since we sinkhole all traffic in P=0.
     per_priority_load[0] = 100;
     return;
   }
 
-  const auto first_available_and_remaining =
+  // We start of with a total load of 100 and distribute it between priorities based on
+  // availability. We first attempt to distribute this load to healthy priorities based on healthy
+  // availability.
+  const auto first_healthy_and_remaining =
       distributeLoad(per_priority_load, per_priority_health, 100, normalized_total_health);
 
-  const size_t remaining_load = first_available_and_remaining.second;
+  // Using the remaining load after allocating load to healthy priorities, distribute it based on
+  // degraded availability.
+  const auto remaining_load_for_degraded = first_healthy_and_remaining.second;
+  const auto first_degraded_and_remaining =
+      distributeLoad(degraded_per_priority_load, per_priority_degraded, remaining_load_for_degraded,
+                     normalized_total_health);
+
+  // Anything that remains should just be rounding errors, so allocate that to the first available
+  // priority, either as healthy or degraded.
+  const auto remaining_load = first_degraded_and_remaining.second;
   if (remaining_load != 0) {
-    ASSERT(first_available_and_remaining.first != -1);
-    // Account for rounding errors by assigning it to the first available priority.
-    ASSERT(remaining_load < per_priority_load.size());
-    per_priority_load[first_available_and_remaining.first] += remaining_load;
+    const auto first_healthy = first_healthy_and_remaining.first;
+    const auto first_degraded = first_degraded_and_remaining.first;
+    ASSERT(first_healthy != -1 || first_degraded != -1);
+
+    // Attempt to allocate the remainder to the first healthy priority first. If no such priority
+    // exist, allocate to the first degraded priority.
+    ASSERT(remaining_load < per_priority_load.size() + per_priority_degraded.size());
+    if (first_healthy != -1) {
+      per_priority_load[first_healthy] += remaining_load;
+    } else {
+      degraded_per_priority_load[first_degraded] += remaining_load;
+    }
   }
+
+  // The allocated load between healthy and degraded should be exactly 100.
+  ASSERT(100 == std::accumulate(per_priority_load.begin(), per_priority_load.end(), 0) +
+                    std::accumulate(degraded_per_priority_load.begin(),
+                                    degraded_per_priority_load.end(), 0));
 }
 
 // Method iterates through priority levels and turns on/off panic mode.
 void LoadBalancerBase::recalculatePerPriorityPanic() {
   per_priority_panic_.resize(priority_set_.hostSetsPerPriority().size());
 
-  const uint32_t normalized_total_health = calcNormalizedTotalHealth(per_priority_health_);
+  const uint32_t normalized_total_health =
+      calcNormalizedTotalHealth(per_priority_health_, per_priority_degraded_);
```

Comment:
Do you think it's worth commenting int he docs that for the purpose of panic calculations that degraded hosts are considered fully unhealthy?

It might otherwise surprise folks that while P=0 degraded hosts are treated as P=1 healthy hosts for load balancing they aren't for panic

Reply:
I'm thinking we might actually want to consider them as healthy wrt panic thresholds since they're still route-able. That said I might leave it as is in this PR to avoid scope creep, so maybe it makes sense to document this for now and follow up with a change in behavior here?

Reply:
Hmm yeah I think this is related to some of my other questions/comments, so some kind of TODO and more comments here would be appreciated.